ParallelPrimitives/RadixSortKernels.h

```diff
@@ -521,278 +514,708 @@ extern "C" __global__ void SortSinglePassKernel( int* gSrcKey, int* gDstKey, int
+__device__ __forceinline__ void onesweep_reorder( RADIX_SORT_KEY_TYPE* inputKeys, RADIX_SORT_KEY_TYPE* outputKeys, RADIX_SORT_VALUE_TYPE* inputValues, RADIX_SORT_VALUE_TYPE* outputValues, bool keyPair, uint32_t numberOfInputs, uint32_t* gpSumBuffer,
+												  volatile uint64_t* lookBackBuffer, uint32_t startBits, uint32_t iteration )
 {
-	const int startOffset = blockIdx.x * gNItemsPerWG;
-	const int nItemInBlock = ( startOffset + gNItemsPerWG > numberOfInputs ) ? numberOfInputs - startOffset : gNItemsPerWG;
-
 	struct ElementLocation
 	{
-		u32 localSrcIndex : 12;
-		u32 localOffset : 12;
-		u32 bucket : 8;
+		uint32_t localSrcIndex : 12;
+		uint32_t localOffset : 12;
+		uint32_t bucket : 8;
 	};
 
-	__shared__ u32 globalOffset[BIN_SIZE];
-	__shared__ u32 localPrefixSum[BIN_SIZE];
-	__shared__ u32 counters[BIN_SIZE];
-	__shared__ u32 matchMasks[SORT_NUM_WARPS_PER_BLOCK][BIN_SIZE];
-	__shared__ u8 elementBuckets[SORT_SUBBLOCK_SIZE];
-	__shared__ ElementLocation elementLocations[SORT_SUBBLOCK_SIZE];
+	__shared__ uint32_t pSum[256];
+	__shared__ uint32_t localPrefixSum[256];
+	__shared__ uint32_t counters[256];
+	__shared__ ElementLocation elementLocations[RADIX_SORT_BLOCK_SIZE];
+	__shared__ uint8_t elementBuckets[RADIX_SORT_BLOCK_SIZE];
+	__shared__ uint32_t matchMasks[REORDER_NUMBER_OF_WARPS][256];
 
-	for( int i = threadIdx.x; i < BIN_SIZE; i += SORT_WG_SIZE )
-	{
-		// Note: The size of gHistogram is always BIN_SIZE * N_WGS_EXECUTED
-		globalOffset[i] = gHistogram[i * N_WGS_EXECUTED + blockIdx.x];
-	}
+	uint32_t bitLocation = startBits + 8 * iteration;
+	uint32_t blockIndex = blockIdx.x;
+	uint32_t numberOfBlocks = div_round_up( numberOfInputs, RADIX_SORT_BLOCK_SIZE );
 
-	for( int w = 0; w < SORT_NUM_WARPS_PER_BLOCK; ++w )
+	clearShared<256, REORDER_NUMBER_OF_THREADS_PER_BLOCK, uint32_t>( localPrefixSum, 0 );
+	clearShared<256, REORDER_NUMBER_OF_THREADS_PER_BLOCK, uint32_t>( counters, 0 );
+
+	for( int w = 0; w < REORDER_NUMBER_OF_WARPS; w++ )
 	{
-		for( int i = threadIdx.x; i < BIN_SIZE; i += SORT_WG_SIZE )
+		for( int i = 0; i < 256; i += REORDER_NUMBER_OF_THREADS_PER_BLOCK )
 		{
-			matchMasks[w][i] = 0;
+			matchMasks[w][i + threadIdx.x] = 0;
 		}
 	}
 
 	__syncthreads();
 
-	for( int j = 0; j < nItemInBlock; j += SORT_SUBBLOCK_SIZE )
+	// count
+#if defined( KEY_IS_16BYTE_ALIGNED )
+	if( ( blockIndex + 1 ) * RADIX_SORT_BLOCK_SIZE <= numberOfInputs )
 	{
-		for( int i = threadIdx.x; i < BIN_SIZE; i += SORT_WG_SIZE )
+		for( int i = 0; i < RADIX_SORT_BLOCK_SIZE; i += REORDER_NUMBER_OF_THREADS_PER_BLOCK * 4 )
 		{
-			counters[i] = 0;
-			localPrefixSum[i] = 0;
+			uint32_t itemIndex = blockIndex * RADIX_SORT_BLOCK_SIZE + i + threadIdx.x * 4;
+			struct alignas( 16 ) Key4
+			{
+				RADIX_SORT_KEY_TYPE xs[4];
+			};
+			Key4 key4 = *(Key4*)&inputKeys[itemIndex];
```

Comment:
This could  potentially lead to reduced performance or a bus error because inputKeys might not be aligned to the boundary of `Key4`.  The `alignas` specifies that every object of type `Key4` will be aligned to 16-byte boundary, but this is not the same for `inputKeys` .

Besides, from my experience and from the experimental results I've conducted previously, converting to this would not make it faster.

Reply:
It's up to us if we enforce users to require 16-byte alignment for input & output. I observed faster execution with global_load_b128 instruction via 16-byte batch loads and that's why I put them.